azurerm/resource_arm_api_management_api_version_set.go

```diff
@@ -0,0 +1,218 @@
+package azurerm
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/helper/validation"
+
+	"github.com/Azure/azure-sdk-for-go/services/apimanagement/mgmt/2018-01-01/apimanagement"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/validate"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmApiManagementApiVersionSet() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmApiManagementApiVersionSetCreateUpdate,
+		Read:   resourceArmApiManagementApiVersionSetRead,
+		Update: resourceArmApiManagementApiVersionSetCreateUpdate,
+		Delete: resourceArmApiManagementApiVersionSetDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		CustomizeDiff: func(diff *schema.ResourceDiff, v interface{}) error {
+			switch schema := diff.Get("versioning_schema").(string); schema {
+			case string(apimanagement.VersioningSchemeSegment):
+				if _, ok := diff.GetOk("version_header_name"); ok {
+					return fmt.Errorf("`version_header_name` can not be set if `versioning_schema` is `Segment`")
+				}
+				if _, ok := diff.GetOk("version_query_name"); ok {
+					return fmt.Errorf("`version_query_name` can not be set if `versioning_schema` is `Segment`")
+				}
+
+			case string(apimanagement.VersioningSchemeHeader):
+				if _, ok := diff.GetOk("version_header_name"); !ok {
+					return fmt.Errorf("`version_header_name` must be set if `versioning_schema` is `Header`")
+				}
+				if _, ok := diff.GetOk("version_query_name"); ok {
+					return fmt.Errorf("`version_query_name` can not be set if `versioning_schema` is `Header`")
+				}
+
+			case string(apimanagement.VersioningSchemeQuery):
+				if _, ok := diff.GetOk("version_query_name"); !ok {
+					return fmt.Errorf("`version_query_name` must be set if `versioning_schema` is `Query`")
+				}
+				if _, ok := diff.GetOk("version_header_name"); ok {
+					return fmt.Errorf("`version_header_name` can not be set if `versioning_schema` is `Query`")
+				}
+			}
+
+			return nil
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": azure.SchemaApiManagementChildName(),
+
+			"resource_group_name": resourceGroupNameSchema(),
+
+			"api_management_name": azure.SchemaApiManagementName(),
+
+			"description": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validate.NoEmptyStrings,
+			},
+
+			"display_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validate.NoEmptyStrings,
+			},
+
+			"versioning_schema": {
```

Comment:
It seems that we can infer this value from HCL. For example, if user provides `header_name`, it would be `Header`. In this way, we do not need to do `CustomizeDiff` check, and the user will not duplicate the effort to match those attributes.

Reply:
that'd be true if `header` and `query` were the only two options - but `segment` being the third option (without a separate field) makes this hard (not at this time, since we can set this if both header & query are set) in the future if a fourth value is added, as such I think it's worth keeping this for the moment?

Reply:
this is `scheme` and not `schema` in the SDK - will push a commit to fix